alacritty/src/input.rs

```diff
@@ -112,252 +111,230 @@ pub trait ActionContext<T: EventListener> {
     fn search_active(&self) -> bool;
     fn on_typing_start(&mut self) {}
     fn toggle_vi_mode(&mut self) {}
+    fn hint_state(&mut self) -> &mut HintState;
+    fn hint_input(&mut self, _character: char) {}
 }
 
-trait Execute<T: EventListener> {
-    fn execute<A: ActionContext<T>>(&self, ctx: &mut A);
-}
-
-impl<T, U: EventListener> Execute<U> for Binding<T> {
-    /// Execute the action associate with this binding.
-    #[inline]
-    fn execute<A: ActionContext<U>>(&self, ctx: &mut A) {
-        self.action.execute(ctx)
-    }
-}
-
-impl Action {
-    fn toggle_selection<T, A>(ctx: &mut A, ty: SelectionType)
-    where
-        A: ActionContext<T>,
-        T: EventListener,
-    {
-        let cursor_point = ctx.terminal().vi_mode_cursor.point;
-        ctx.toggle_selection(ty, cursor_point, Side::Left);
+macro_rules! execute_action {
```

Comment:
What is the reason behind putting everything into macros here?

Reply:
Macros allow calling this without being forced to clone the bindings. I kinda hate it but I don't see any other solution at the moment. Generally since I've started using an `Rc<Cell<T>>` for the regex it's not strictly necessary anymore, but it still saves us a `clone` of the binding when trying to trigger it.

Fundamentally the problem is that triggering a `binding.execute(&mut ctx)` causes issues due to the mutable access with an existing reference. With a macro and everything inlined this of course isn't a problem because the stuff can all be resolved inside the current function, allowing for more dynamic lifetime resolution by dropping things early.

Reply:
I've removed the macro since the cloning was already present before this PR anyways. No need to change it right now.